### PR TITLE
feat: FM22xx multi-channel support + alarm target read/write

### PR DESCRIPTION
**Pull request by nberenbold**
## Summary

This PR adds full support for the FM22xx series (e.g. FM2201+) alongside the existing FMC series support.

### New features
- **Multi-channel parsing**: Channel 0 = FMC (original formula), Channel 1 = FM22xx probe 1, Channel 2+ = FM22xx probe 2
- **`probe_battery_2`**: Second probe battery from channel 2 frame
- **`ambient_temperature` / `ambient_temperature_2`**: Per-channel ambient temperature from bytes 6–7 of each probe frame (verified: CH1=27°C, CH2=30°C simultaneously)
- **Alarm target read/write**: `write_target(ch, temp_c)` and `clear_target(ch)` for FM22xx (protocol: `01 03 [CH] [lo] [hi]` / `02 03 [CH]`)
- **STATUS frame (0x00 0x05, 41 B)**: Reads current alarm targets at init
- **TARGET frame (0x00 0x03)**: Updates targets on change

### Bug fixes
- **Relay voltage byte order**: FM22xx uses little-endian, FMC uses big-endian. Was unconditionally big-endian → 12.3V instead of ~3.9V → station battery always 100%. Fixed: `endian = 'little' if self._is_fm22 else 'big'`. Verified via BLE sniffer: `0x30 0x0f` LE = 3888 mV → 74%.
- **RSSI signed byte**: `data[8]` is signed dBm (e.g. `0xD7` = −41 dBm, not 215). Fixed with `int.from_bytes([data[8]], signed=True)`.
- **`_is_fm22` detection**: Flag is set to `True` in STATUS and TARGET frames (FM22xx-only), ensuring correct relay thresholds even if relay frame arrives before the first probe frame. Set to `False` for channel 0 (FMC) frames.

### Input validation
- `write_target` / `clear_target`: channel must be 1 or 2; temperature encoding validated against 0xFFFF sentinel.

### Version
1.2.0 → 1.3.0

---

## How the values were found

All values in this PR were derived through a combination of OEM APK decompilation and live BLE frame sniffing — no guesswork, no OTA writes.

### Step 1 — APK extraction

The OEM app for the FM2201+ is `com.easycheck.thermoex` (rebranded by Outdoorchef). It was pulled directly from an Android device via ADB:

```
adb shell pm path com.easycheck.thermoex
adb pull <path>/base.apk
```

### Step 2 — Decompilation with jadx

Since Java is not available on the development machine, jadx 1.5.5 was run inside a Docker container:

```bash
docker run --rm \
  -v ~/jadx/libexec:/jadx:ro \
  -v /tmp/thermoex:/work \
  eclipse-temurin:17 \
  /jadx/bin/jadx --show-bad-code -d /work/out /work/base.apk
```

The `--show-bad-code` flag was needed because jadx could not fully decompile `onCharacteristicChanged` — the flag produces pseudo-code for unresolvable sections.

### Step 3 — Static analysis

The relevant class is `com.yscoco.fsdapp.other.ble.Ble2Utils`, specifically the `AnonymousClass6` BLE notify callback. Key findings:

**Station battery thresholds** (from `DeviceDao.setBattery` call chain in the relay frame handler):
```java
if (iHexString2Int >= 3900) setBattery(100)
else if (iHexString2Int >= 3700) setBattery(74)
else if (iHexString2Int >= 3460) setBattery(49)
else setBattery(0)
```
These are raw mV integers read from the relay frame.

**Oven/ambient temperature** (bytes 6–7 of the probe frame, labelled 炉温 "furnace temperature" in the source):
```java
// channel frame bytes 6-7: little-endian int16 / 10 = °C
```

**RSSI** (byte 8 of the probe frame): signed dBm, cast explicitly as signed in the OEM source.

**Device family detection**: The class `DeviceModel.FM2` covers all FM22xx devices. Channel byte `data[2] != 0x00` distinguishes FM22xx from FMC single-channel devices.

### Step 4 — Live BLE verification

A Python sniffer script using `bleak` was run directly against the FM2201+ hardware to verify every parsed value:

```
NOTIFY [ff01] len=8  00 01 30 0f 01 01 ff ff
→ bytes[2:4] = 0x30 0x0f
   big-endian:    0x300f = 12303 mV → always 100% (wrong)
   little-endian: 0x0f30 =  3888 mV → 74%          (correct, matches OEM app)

NOTIFY [ff01] len=9  00 00 01 64 00 01 0e 01 d7
→ CH1: temp=25.6°C  ambient=27.0°C  probe_v=3.125V  rssi=0xD7=−41 dBm

NOTIFY [ff01] len=9  00 00 02 64 00 01 12 01 d7
→ CH2: temp=25.6°C  ambient=30.0°C  (different ambient per channel confirmed)
```

This confirmed the little-endian bug in the relay frame and proved that each probe channel carries its own independent ambient temperature reading.

## Test plan
- [x] BLE sniffer confirms relay_voltage ~3.888 V → 74% station battery (FM2201+)
- [x] CH1 and CH2 report separate ambient temperatures (27°C vs 30°C)
- [x] RSSI shows correct negative dBm values
- [x] Station battery no longer stuck at 100%

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by nberenbold** [comment]
@pantherale0  Sorry for the AI Slop but I didn't had the time to personally investigate all the addresses and stuff but wanted my outdoor chef probe to work.
PR for HA will follow soon

File: pyprobeplus/__init__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 from __future__ import annotations
 
-__version__ = "1.0.1"
```
Comment (pantherale0) [inline pyprobeplus/__init__.py:5]: Version string should not be changed - this is handled automatically by the CI workflow

File: pyprobeplus/__init__.py [hunk 2]
```diff
@@ -29,6 +29,7 @@ def __init__(
         scanner: BleakScanner | None = None,
         name: str | None = None,
         notify_callback: Callable[[], None] | None = None,
+        is_fm22: bool = False,
```
Comment (pantherale0) [inline pyprobeplus/__init__.py]: Is there a way we can figure this out dynamically? Perhaps by the name of the device?

File: pyprobeplus/parser.py
Comment (pantherale0) [inline pyprobeplus/parser.py:1]: Will review this later, because I think there should just be a base class with overrides for different devices

**Review by pantherale0 — CHANGES_REQUESTED** [review]
Thanks, just a couple of initial comments :)

File: pyprobeplus/parser.py
Comment (nberenbold) [reply pyprobeplus/parser.py:1]: If you have an idea how this can be intelligent done, just give me a headsup, right now I just wanted to get it to work.

File: pyprobeplus/__init__.py (hunk 2 above)
Comment (nberenbold) [reply pyprobeplus/__init__.py]: This could potentially collide since there are now many ones of them, but let me investigate once I have the time.
Also perhaps some of the changes which are done here are applicable for the old integration like Station Battery and such

**Comment by pantherale0** [comment]
I just wanted to check-in - I haven't entirely forgotten about this, I think I need to decompile the app again to find the functions and data parsers for different devices. Its clear that it is somehow able to detect differences and use variations of the same protocol depending on what device is connected (#10), but this needs to be figured out first.

File: pyprobeplus/__init__.py (hunk 1 above)
Comment (nberenbold) [reply pyprobeplus/__init__.py:5]: Reverted — kept at 1.0.1 (the last value before this PR touched it). Left `__version__` alone from here on; the release workflow's `sed` sets it from the tag.

File: pyprobeplus/__init__.py (hunk 2 above)
Comment (nberenbold) [reply pyprobeplus/__init__.py]: Done — added `parser_for_device(name)` which picks the parser class from the advertised BLE name (any name starting with "FM2" -> the FM2 parser, everything else -> the original FMC parser). The `is_fm22` constructor kwarg is gone entirely; `ProbePlusDevice` resolves the name from either the explicit `name` arg or a passed-in `BLEDevice`, and re-resolves it in `connect()` if it wasn't known yet at construction time (covered by tests in test/test_device.py — this needed two follow-up fixes after automated review caught edge cases: a bare MAC-string with no name, and a BLEDevice whose name isn't populated yet).

File: pyprobeplus/parser.py
Comment (nberenbold) [reply pyprobeplus/parser.py:1]: Done — `ParserBase` now holds the original FMC decode unchanged, and a new `Fm2Parser(ParserBase)` overrides the device-specific pieces (temperature format, relay endianness/thresholds, channel-to-probe-slot mapping, ambient temperature, STATUS/TARGET alarm frames) instead of branching inline on a runtime flag. `ProbePlusData` also moved from flat per-probe fields to a `probes: list[ProbeReading]`, with the old flat fields kept as read-only properties mirroring `probes[0]`/`probes[1]` for backward compatibility. Also fixes a real bug this surfaced: channel-based family detection was decoding FM210+/INSMART devices (github issue #10) with the wrong (FMC) formula, since they send channel 0 just like FMC — name-based detection handles the whole FM2 family (FM2201+, FM2209, FM210+) uniformly.

**Comment by nberenbold** [comment]
Addressed all three review comments above (replies inline on each). Summary:

- `__version__` reverted to 1.0.1, left alone going forward.
- `is_fm22` constructor kwarg removed entirely — `ProbePlusDevice` now resolves the parser from the advertised BLE name (`parser_for_device(name)`), re-resolving in `connect()` if the name wasn't known yet at construction.
- `ParserBase` (FMC, unchanged) + `Fm2Parser` subclass (FM2201+/FM2209/FM210+/INSMART) with overridable hooks instead of the `_is_fm22`-flag branching.
- `ProbePlusData.probes: list[ProbeReading]` as the canonical per-probe model; old flat fields (`probe_temperature_2`, `ambient_temperature_2`, `target_1`/`target_2`, ...) kept as read-only properties mirroring `probes[0]`/`probes[1]`, so nothing downstream breaks.
- Fixes a real bug the channel-based detection had: FM210+/INSMART (issue #10) sends channel 0 like FMC and was getting decoded with the wrong formula — name-based detection handles the whole FM2 family the same way, so this PR now also resolves #10.
- Also fixed along the way (own testing + two rounds of independent automated review — Codex/CodeRabbit — on the intermediate branch): `probe_voltage`/`probe_rssi` being overwritten across channels, the relay-frame endianness race, `int.from_bytes` needing `signed=True` with an explicit byteorder for py3.8 compat, a phantom second probe being created on single-probe devices when a STATUS/TARGET frame arrives, and STATUS/TARGET data arriving before a probe's first frame no longer being lost (now applied once that probe is seen).
- `test/test_parser.py` + `test/test_device.py` added (the repo had no tests before) — 32 tests, `parser.py` at 100% line+branch coverage. Fixtures are real BLE frames from this PR's own sniffing and from issue #10.

One open design question I didn't want to decide unilaterally: `ProbePlusData(probe_temperature=25.0)`-style direct keyword construction of the legacy flat fields no longer works (they're derived read-only properties now). Nothing in this repo or my tests constructs it that way, but it is a change to the public dataclass constructor. Happy to add a compat shim if you'd rather keep it.

Also tested end-to-end against real FM2201+ hardware on my own Home Assistant instance (vendored copy) — both probes decode independently, station battery no longer stuck at 100%, alarm target write/clear round-trips over BLE correctly.

**Comment by nberenbold** [comment]
> I just wanted to check-in - I haven't entirely forgotten about this, I think I need to decompile the app again to find the functions and data parsers for different devices. Its clear that it is somehow able to detect differences and use variations of the same protocol depending on what device is connected (#10), but this needs to be figured out first.

Sorry had a lot todo but found time today to tackle this or let it being tackled, perhaps you have some time to review it later
